### PR TITLE
Refactors incoming and outgoing message-related metrics

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR implements the following changes:
1. Introducing `IncomingMessageScope` and `OutgoingMessageScope` as wrappers around `message.Message` type which cuts the dependencies to its deprecated fields and unifies the interface across the networking layer. 

2. Splitting the `metrics.NetworkMetrics` into smaller abstractions hence restricting the access of networking components only to the metrics that they need. 

3. Consolidating the metrics collection for the core networking components into the `p2p.Network` component.

4. Updating the names of obsolete variables, methods and functions.

### CODE REVIEW

File: network/internal/messageutils/utils.go [hunk 1]
```diff
@@ -1,49 +0,0 @@
...
-
-	"github.com/onflow/flow-go/model/flow"
-	libp2pmessage "github.com/onflow/flow-go/model/libp2p/message"
-	"github.com/onflow/flow-go/network/channels"
-	"github.com/onflow/flow-go/network/message"
-	"github.com/onflow/flow-go/network/p2p"
-	"github.com/onflow/flow-go/utils/unittest"
-)
-
-func CreateMessage(t *testing.T, originID flow.Identifier, targetID flow.Identifier, channel channels.Channel, msg string) (*message.Message, *message.Message, interface{}) {
```
Comment (yhassanzadeh13) [inline network/internal/messageutils/utils.go:16]: With the changes in this PR, we no longer deal directly with the `message.Message` in testing. Rather, we use the `IncomingMessageScope` and `OutgoingMessageScope`.

File: network/message/message.pb.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ type Message struct {
 	OriginID             []byte   `protobuf:"bytes,3,opt,name=OriginID,proto3" json:"OriginID,omitempty"` // Deprecated: Do not use.
 	TargetIDs            [][]byte `protobuf:"bytes,4,rep,name=TargetIDs,proto3" json:"TargetIDs,omitempty"`
 	Payload              []byte   `protobuf:"bytes,5,opt,name=Payload,proto3" json:"Payload,omitempty"`
-	Type                 string   `protobuf:"bytes,6,opt,name=Type,proto3" json:"Type,omitempty"` // Deprecated: Do not use.
+	Type                 string   `protobuf:"bytes,6,opt,name=Protocol,proto3" json:"Protocol,omitempty"` // Deprecated: Do not use.
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/message/message.pb.go:33]: you are right, it was not intended, I am reverting it back.

File: network/p2p/middleware/readSubscription.go [hunk 3]
```diff
@@ -21,49 +20,36 @@ import (
 type readSubscriptionCB func(msg *message.Message, decodedMsgPayload interface{}, peerID peer.ID)
 
 // readSubscription reads the messages coming in on the subscription and calls the given callback until
-// the context of the subscription is cancelled. Additionally, it reports metrics
+// the context of the subscription is cancelled.
 type readSubscription struct {
-	ctx      context.Context
 	log      zerolog.Logger
 	sub      p2p.Subscription
-	metrics  module.NetworkMetrics
```
Comment (yhassanzadeh13) [inline network/p2p/middleware/readSubscription.go:29]: We consolidate all metrics collections into the `p2p.Network`.

File: network/p2p/middleware/middleware.go [hunk 4]
```diff
@@ -685,24 +678,22 @@ func (m *Middleware) processUnicastStreamMessage(remotePeer peer.ID, msg *messag
...
-				Str("type", msg.Type).
+				Str("type", messageType).
 				Str("channel", msg.ChannelID).
 				Msg("unicast authorized sender validation failed")
 			return
 		}
 	}
 
-	// message decoding and validation was successful now log metrics with the channel name as OneToOne and process message
-	m.metrics.NetworkMessageReceived(msg.Size(), metrics.ChannelOneToOne, msg.Type)
```
Comment (yhassanzadeh13) [inline network/p2p/middleware/middleware.go:696]: Consolidating all metrics into `p2p.Network`.

File: network/p2p/unicast/ratelimit/rate_limiters.go [hunk 5]
```diff
@@ -80,13 +79,13 @@ func (r *RateLimiters) MessageAllowed(peerID peer.ID) bool {
 
 // BandwidthAllowed will return result from BandWidthRateLimiter.Allow. It will invoke the OnRateLimitedPeerFunc
 // callback each time a peer is not allowed.
-func (r *RateLimiters) BandwidthAllowed(peerID peer.ID, role string, msg *message.Message) bool {
```
Comment (yhassanzadeh13) [inline network/p2p/unicast/ratelimit/rate_limiters.go:83]: sticking to the Least Privilege Principle, let's only feed the information that this component needs.

File: network/validator/sender_validator.go [hunk 6]
```diff
@@ -1,30 +1,29 @@
...
 	"github.com/onflow/flow-go/network"
-	"github.com/onflow/flow-go/network/message"
 )
 
 var _ network.MessageValidator = &SenderValidator{}
 
 // SenderValidator validates messages by sender ID
 type SenderValidator struct {
-	sender []byte
+	sender flow.Identifier
```
Comment (yhassanzadeh13) [inline network/validator/sender_validator.go:12]: With message scopes defined, we can now fully rely on the `flow.Identifier`s.

File: network/p2p/rate_limiter.go [hunk 7]
```diff
@@ -4,14 +4,12 @@ import (
...
-
-	"github.com/onflow/flow-go/network/message"
 )
 
 // RateLimiter unicast rate limiter interface
 type RateLimiter interface {
-	// Allow returns true if a message should be allowed to be processed.
-	Allow(peerID peer.ID, msg *message.Message) bool
+	// Allow returns true if a message with the give size should be allowed to be processed.
+	Allow(peerID peer.ID, msgSize int) bool
```
Comment (yhassanzadeh13) [inline network/p2p/rate_limiter.go:12]: Let's stick to the Least Privilege Principal: `Allow` seems to only need to the message size.

File: network/p2p/network.go [hunk 8]
```diff
@@ -524,23 +506,3 @@ func (n *Network) queueSubmitFunc(message interface{}) {
...
-	if err != nil {
-		return nil, fmt.Errorf("could not hash event: %w", err)
-	}
-
-	return h.SumHash(), nil
-}
-
-func MessageType(decodedPayload interface{}) string {
-	return strings.TrimLeft(fmt.Sprintf("%T", decodedPayload), "*")
-}
```
Comment (yhassanzadeh13) [inline network/p2p/network.go:546]: moved to `network` package.

File: network/message_scope.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+func (m IncomingMessageScope) DecodedPayload() interface{} {
+	return m.decodedPayload
+}
+
+func (m IncomingMessageScope) Protocol() ProtocolType {
+	return m.protocol
+}
+
+func (m IncomingMessageScope) Channel() string {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/message_scope.go]: https://github.com/onflow/flow-go/pull/3675/commits/132fe2c9888504ff2e716a8e90048456669449ea

File: network/p2p/middleware/middleware.go [hunk 10]
```diff
@@ -714,41 +705,32 @@ func (m *Middleware) processAuthenticatedMessage(msg *message.Message, decodedMs
 		return
 	}
 
-	channel := channels.Channel(msg.ChannelID)
-	eventID, err := p2p.EventId(channel, msg.GetPayload())
+	scope, err := network.NewIncomingScope(originId, protocol, msg, decodedMsgPayload)
 	if err != nil {
 		m.log.Error().
 			Err(err).
 			Str("peer_id", peerID.String()).
```
Comment (kc1116) [inline network/p2p/middleware/middleware.go:712]:
```suggestion
			Str("peer_id", peerID.String()).
			Str("origin_id", originId.String()).
```

File: network/test/middleware_test.go [hunk 11]
```diff
@@ -1,12 +1,16 @@
...
 import (
+	"bytes"
 	"context"
 	"fmt"
+	"regexp"
 	"sync"
 	"testing"
 	"time"
 
+	"github.com/onflow/flow-go/network/p2p/middleware"
```
Comment (kc1116) [inline network/test/middleware_test.go:12]: nit: group import with the rest of the flow imports

File: network/test/middleware_test.go [hunk 12]
```diff
@@ -431,38 +462,6 @@ func (m *MiddlewareTestSuite) createOverlay(provider *testutils.UpdatableIDProvi
 	return overlay
 }
 
-// TestPingRawReception tests the middleware for solely the
```
Comment (kc1116) [inline network/test/middleware_test.go:434]: Why are these tests not needed anymore ?

File: network/validator/target_validator.go [hunk 13]
```diff
@@ -1,39 +1,43 @@
 package validator
 
 import (
-	"bytes"
+	"github.com/onflow/flow-go/utils/logging"
```
Comment (kc1116) [inline network/validator/target_validator.go:4]: nit: fix imports

**Review by kc1116 — COMMENTED** [review]
Nice refactor everything looks good, left a few small comments.

File: network/test/middleware_test.go [hunk 14]
```diff
@@ -509,33 +525,63 @@ func (m *MiddlewareTestSuite) Ping(msg *message.Message, expectID, expectedMessa
...
+
+				msg, ok := args[0].(*network.IncomingMessageScope)
+				require.True(m.T(), ok)
+
+				require.Equal(m.T(), testChannel, msg.ChannelId())                    // channel
+				require.Equal(m.T(), m.ids[firstNodeIndex].NodeID, msg.OriginId())    // sender id
+				require.Equal(m.T(), m.ids[lastNodeIndex].NodeID, msg.TargetIDs()[0]) // target id
+				require.Equal(m.T(), network.ProtocolTypeUnicast, msg.Protocol())     // protocol
+
+				mu.Lock() // protects receivedPayloads
```
Comment (peterargue) [inline network/test/middleware_test.go:572]: is this to protect concurrent access to `receivedPayloads`? You might consider using [ProtectedMap](https://github.com/onflow/flow-go/blob/master/utils/unittest/protected_map.go#L6-L6)

File: network/message_scope.go [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+type OutgoingMessageScope struct {
+	targetIds flow.IdentifierList               // the target node IDs.
+	channelId channels.Channel                  // the channel ID.
+	payload   interface{}                       // the payload to be sent.
+	encoder   func(interface{}) ([]byte, error) // the encoder to encode the payload.
+	msg       *message.Message                  // raw proto message sent on wire.
+	protocol  ProtocolType                      // the type of protocol used to send the message.
+}
+
+func NewOutgoingScope(
```
Comment (peterargue) [inline network/message_scope.go:111]: please document the return errors

File: network/message_scope.go [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+		Payload:   payload,
+	}, nil
+}
+
+func (o OutgoingMessageScope) Proto() *message.Message {
+	return o.msg
+}
+
+// EventId computes the event ID for a given channel and payload (i.e., the hash of the payload and channel).
+func EventId(channel channels.Channel, payload []byte) (hash.Hash, error) {
```
Comment (peterargue) [inline network/message_scope.go:187]: please document the return errors

File: network/message_scope.go [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+type IncomingMessageScope struct {
+	originId       flow.Identifier     // the origin node ID.
+	targetIds      flow.IdentifierList // the target node IDs (i.e., intended recipients).
+	eventId        hash.Hash           // hash of the payload and channel.
+	msg            *message.Message    // the raw message received.
+	decodedPayload interface{}         // decoded payload of the message.
+	protocol       ProtocolType        // the type of protocol used to receive the message.
+}
+
+func NewIncomingScope(originId flow.Identifier, protocol ProtocolType, msg *message.Message, decodedPayload interface{}) (*IncomingMessageScope, error) {
```
Comment (peterargue) [inline network/message_scope.go:45]: please document the return errors

**Review by peterargue — COMMENTED** [review]
looking pretty good. Added a few small comments

File: network/test/middleware_test.go (hunk 11 above)
Comment (yhassanzadeh13) [reply network/test/middleware_test.go]: https://github.com/onflow/flow-go/pull/3675/commits/c91c88ec46b9cf835ac81dacb279b90cbfb397bd

File: network/validator/target_validator.go (hunk 13 above)
Comment (yhassanzadeh13) [reply network/validator/target_validator.go]: https://github.com/onflow/flow-go/pull/3675/commits/82e0838d40257bdab961628c6098a3e40ca045d0

File: network/test/middleware_test.go (hunk 12 above)
Comment (yhassanzadeh13) [reply network/test/middleware_test.go:434]: These four tests have been consolidated into the `TestPing`:
- `TestPingRawReception` tests payload of the received message, and `TestPing` now covers this evaluation [here](https://github.com/onflow/flow-go/pull/3675/files/9087a41417c995b2604600123ae027503859c407#diff-199e0c5e2e32815e4ac1058c9c089d4a9f8a7059f6399145c6156445b7bb3642R512).
- `TestPingTypeReception` tests type of the received payload, and `TestPing` now covers this evaluation [here](https://github.com/onflow/flow-go/pull/3675/files/9087a41417c995b2604600123ae027503859c407#diff-199e0c5e2e32815e4ac1058c9c089d4a9f8a7059f6399145c6156445b7bb3642R512). 
- `TestPingIDType` tests the origin id of the received message, and we have an enhanced testing of this (both type and value) [here](https://github.com/onflow/flow-go/pull/3675/files/9087a41417c995b2604600123ae027503859c407#diff-199e0c5e2e32815e4ac1058c9c089d4a9f8a7059f6399145c6156445b7bb3642R509) in `TestPing`. 
- `TestPingContentReceiption` tests the entire scope of the received message, which is now covered by `TestPing` [here](https://github.com/onflow/flow-go/pull/3675/files/9087a41417c995b2604600123ae027503859c407#diff-199e0c5e2e32815e4ac1058c9c089d4a9f8a7059f6399145c6156445b7bb3642R503-R512).

File: network/message_scope.go [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+
+func (m IncomingMessageScope) DecodedPayload() interface{} {
+	return m.decodedPayload
+}
+
+func (m IncomingMessageScope) Protocol() ProtocolType {
+	return m.protocol
+}
+
+func (m IncomingMessageScope) ChannelId() channels.Channel {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/message_scope.go]: https://github.com/onflow/flow-go/pull/3675/commits/9f2ec6c790282b271af6a5fa67bb21fa4b37c644

File: network/p2p/middleware/middleware.go [hunk 19]
```diff
@@ -790,16 +775,14 @@ func (m *Middleware) Publish(msg *message.Message, channel channels.Channel) err
 		return fmt.Errorf("message size %d exceeds configured max message size %d", msgSize, p2pnode.DefaultMaxPubSubMsgSize)
 	}
 
-	topic := channels.TopicFromChannel(channel, m.rootBlockID)
+	topic := channels.TopicFromChannel(channels.Channel(msg.ChannelId()), m.rootBlockID)
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/middleware/middleware.go]: https://github.com/onflow/flow-go/pull/3675/commits/74e4eac555258f7dadaa5c309c6547d25165c719

File: network/test/middleware_test.go (hunk 14 above)
Comment (yhassanzadeh13) [reply network/test/middleware_test.go]: https://github.com/onflow/flow-go/pull/3675/commits/350f469a64469f7d888e732cae784e6c26d5eb44

File: network/message_scope.go (hunk 15 above)
Comment (yhassanzadeh13) [reply network/message_scope.go:116]: https://github.com/onflow/flow-go/pull/3675/commits/f18504204c7df509d213cdc1e67cacf670161e04

File: network/message_scope.go (hunk 16 above)
Comment (yhassanzadeh13) [reply network/message_scope.go:194]: https://github.com/onflow/flow-go/pull/3675/commits/48c62f6999b464bec1742dca7fff9acba6344d9c

File: network/message_scope.go (hunk 17 above)
Comment (yhassanzadeh13) [reply network/message_scope.go:50]: https://github.com/onflow/flow-go/pull/3675/commits/f6d39a051221eedf3b2bfc42a90ff21d9a546369

**Review by kc1116 — APPROVED** [review]

**Review by peterargue — APPROVED** [review]

**Pull request merged**